### ISSUE TITLE
failed to create process

### ISSUE THREAD

**Issue by marcusobrien** [facebookresearch/llama#694]
Hi, 

I am using Windows 10, with a conda env, and have CUDA and everything else setup and working as per the instructions. When I try to run my first example I get a failed to start process

`
(llama2env) H:\llama2\repo\llama\llama>torchrun --nproc_per_node 1 example_text_completion.py \

failed to create process.
`

I can't even get to type in the second line of the script, or if I put the whole thing on one line in my cmd.exe 

torchrun --nproc_per_node 1 example_text_completion.py --ckpt_dir llama-2-7b --tokenizer_path tokenizer.model --max_seq_len 128 --max_batch_size 4

I still get the same error

**failed to create process.**

I can run the tests to get back a tensor

`
import torch
x = torch.rand(5, 3)
print(x)

tensor([[0.5495, 0.0281, 0.2566],
        [0.7032, 0.1296, 0.8173],
        [0.0329, 0.5500, 0.3025],
        [0.6790, 0.0561, 0.3389],
        [0.4403, 0.5365, 0.5513]])
`

And also - ensure CUDA and my 4090 GPU is available

>>> torch.cuda.is_available()
True

The output from **conda info**

`
active environment : llama2env
    active env location : U:\Miniconda3\envs\llama2env
            shell level : 1
       user config file : C:\Users\marcu\.condarc
 populated config files : C:\Users\marcu\.condarc
          conda version : 22.9.0
    conda-build version : not installed
         python version : 3.8.17.final.0
       virtual packages : __cuda=12.2=0
                          __win=0=0
                          __archspec=1=x86_64
       base environment : U:\Miniconda3  (writable)
      conda av data dir : U:\Miniconda3\etc\conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/win-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/win-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/msys2/win-64
                          https://repo.anaconda.com/pkgs/msys2/noarch
          package cache : U:\Miniconda3\pkgs
                          C:\Users\marcu\.conda\pkgs
                          C:\Users\marcu\AppData\Local\conda\conda\pkgs
       envs directories : U:\Miniconda3\envs
                          C:\Users\marcu\.conda\envs
                          C:\Users\marcu\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/22.9.0 requests/2.31.0 CPython/3.8.17 Windows/10 Windows/10.0.19045
          administrator : False
             netrc file : None
           offline mode : False
`

Python 3.11.4 version installed in this conda environment

(llama2env) H:\llama2\repo\llama\llama>python
Python 3.11.4 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 13:47:18) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.

And the list of the Python 3.11.4 packages installed in this conda env

`(llama2env) H:\llama2\repo\llama\llama>conda list
# packages in environment at U:\Miniconda3\envs\llama2env:
#
# Name                    Version                   Build  Channel
blas                      1.0                         mkl
brotlipy                  0.7.0           py311h2bbff1b_1002
bzip2                     1.0.8                he774522_0
ca-certificates           2023.05.30           haa95532_0
certifi                   2023.7.22       py311haa95532_0
cffi                      1.15.1          py311h2bbff1b_3
charset-normalizer        2.0.4              pyhd3eb1b0_0
cryptography              41.0.2          py311hac1b9e3_0
cuda-cccl                 12.2.128                      0    nvidia
cuda-cudart               11.8.89                       0    nvidia
cuda-cudart-dev           11.8.89                       0    nvidia
cuda-cupti                11.8.87                       0    nvidia
cuda-libraries            11.8.0                        0    nvidia
cuda-libraries-dev        11.8.0                        0    nvidia
cuda-nvrtc                11.8.89                       0    nvidia
cuda-nvrtc-dev            11.8.89                       0    nvidia
cuda-nvtx                 11.8.86                       0    nvidia
cuda-profiler-api         12.2.128                      0    nvidia
cuda-runtime              11.8.0                        0    nvidia
fairscale                 0.4.13                   pypi_0    pypi
filelock                  3.9.0           py311haa95532_0
fire                      0.5.0                    pypi_0    pypi
freetype                  2.12.1               ha860e81_0
giflib                    5.2.1                h8cc25b3_3
idna                      3.4             py311haa95532_0
intel-openmp              2023.1.0         h59b6b97_46319
jinja2                    3.1.2           py311haa95532_0
jpeg                      9e                   h2bbff1b_1
lerc                      3.0                  hd77b12b_0
libcublas                 11.11.3.6                     0    nvidia
libcublas-dev             11.11.3.6                     0    nvidia
libcufft                  10.9.0.58                     0    nvidia
libcufft-dev              10.9.0.58                     0    nvidia
libcurand                 10.3.3.129                    0    nvidia
libcurand-dev             10.3.3.129                    0    nvidia
libcusolver               11.4.1.48                     0    nvidia
libcusolver-dev           11.4.1.48                     0    nvidia
libcusparse               11.7.5.86                     0    nvidia
libcusparse-dev           11.7.5.86                     0    nvidia
libdeflate                1.17                 h2bbff1b_0
libffi                    3.4.4                hd77b12b_0
libnpp                    11.8.0.86                     0    nvidia
libnpp-dev                11.8.0.86                     0    nvidia
libnvjpeg                 11.9.0.86                     0    nvidia
libnvjpeg-dev             11.9.0.86                     0    nvidia
libpng                    1.6.39               h8cc25b3_0
libtiff                   4.5.0                h6c2663c_2
libuv                     1.44.2               h2bbff1b_0
libwebp                   1.2.4                hbc33d0d_1
libwebp-base              1.2.4                h2bbff1b_1
llama                     0.0.1                     dev_0    <develop>
lz4-c                     1.9.4                h2bbff1b_0
markupsafe                2.1.1           py311h2bbff1b_0
mkl                       2023.1.0         h6b88ed4_46357
mkl-service               2.4.0           py311h2bbff1b_1
mkl_fft                   1.3.6           py311hf62ec03_1
mkl_random                1.2.2           py311hf62ec03_1
mpmath                    1.3.0           py311haa95532_0
networkx                  3.1             py311haa95532_0
numpy                     1.25.2          py311hdab7c0b_0
numpy-base                1.25.2          py311hd01c5d8_0
openssl                   3.0.10               h2bbff1b_0
pillow                    9.4.0           py311hd77b12b_0
pip                       23.2.1          py311haa95532_0
pycparser                 2.21               pyhd3eb1b0_0
pyopenssl                 23.2.0          py311haa95532_0
pysocks                   1.7.1           py311haa95532_0
python                    3.11.4               he1021f5_0
pytorch                   2.0.1           py3.11_cuda11.8_cudnn8_0    pytorch
pytorch-cuda              11.8                 h24eeafa_5    pytorch
pytorch-mutex             1.0                        cuda    pytorch
requests                  2.31.0          py311haa95532_0
sentencepiece             0.1.99                   pypi_0    pypi
setuptools                68.0.0          py311haa95532_0
six                       1.16.0                   pypi_0    pypi
sqlite                    3.41.2               h2bbff1b_0
sympy                     1.11.1          py311haa95532_0
tbb                       2021.8.0             h59b6b97_0
termcolor                 2.3.0                    pypi_0    pypi
tk                        8.6.12               h2bbff1b_0
torchaudio                2.0.2                    pypi_0    pypi
torchvision               0.15.2                   pypi_0    pypi
typing_extensions         4.7.1           py311haa95532_0
tzdata                    2023c                h04d1e81_0
urllib3                   1.26.16         py311haa95532_0
vc                        14.2                 h21ff451_1
vs2015_runtime            14.27.29016          h5e58377_2
wheel                     0.38.4          py311haa95532_0
win_inet_pton             1.1.0           py311haa95532_0
xz                        5.4.2                h8cc25b3_0
zlib                      1.2.13               h8cc25b3_0
zstd                      1.5.5                hd43e919_0`

**Comment by marcusobrien** [comment]
By the way when I use a normal python environment without conda, the above works perfectly. So its a torchrun with conda issue.

**Comment by marcusobrien** [comment]
The torchrun.exe does not work at all for me. I dont know if its a windows pro setup thing or not, but being an executable (.exe) it is definetly a windows file and therefore should at least execute - the only thing I can think of is its an exe built for a different windows arch ie not x86 ? Running the script **torchrun-script** works.

**Comment by haozhuang0000** [comment]
Hi There, did you solve this issue? i have the same issue. "failed to create process"

**Comment by suyvv** [comment]
You need to edit the torchrun-script.py first line, change it's path to match your python.exe in your conda environment, mine looks like this: #!e:\python_env\collama\python.exe  as I've setup venv in this "e:\python_env\collama"

**Comment by wh1tew0lf0fr1v1a** [comment]
> You need to edit the torchrun-script.py first line, change it's path to match your python.exe in your conda environment, mine looks like this: #!e:\python_env\collama\python.exe as I've setup venv in this "e:\python_env\collama"

were is torchrun-script.py ?

**Comment by musclechen** [comment]
I'm not sure if this is my setting problem.
I modified the 'torchrun-script.py' and then it looks like it works. But I saw a line.
![image](https://github.com/facebookresearch/llama/assets/16273693/ee490a66-dea8-4507-a299-ac66462491fb)
So, maybe this model is not working under Windows. Because I use my server PC(CentOS) it is working.

**Comment by kgopfa** [comment]
> I'm not sure if this is my setting problem. I modified the 'torchrun-script.py' and then it looks like it works. But I saw a line. ![image](https://private-user-images.githubusercontent.com/16273693/289196293-ee490a66-dea8-4507-a299-ac66462491fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxMjQ3ODEsIm5iZiI6MTcwMjEyNDQ4MSwicGF0aCI6Ii8xNjI3MzY5My8yODkxOTYyOTMtZWU0OTBhNjYtZGVhOC00NTA3LWEyOTktYWM2NjQ2MjQ5MWZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA5VDEyMjEyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyYTBlYjhjNjljMmUyNjNmMjFjZDhlZGYxZTZjMWYzOTExZTYzZjBmZTVhMGE3NWY0MTZhMDQxNDdjYTEwOWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.X8Vo24bqQf71rm5IorITKmZ86bjv6opGw4RbfJGRm3g) So, maybe this model is not working under Windows. Because I use my server PC(CentOS) it is working.

I am getting the exact issue, I also tried running the model on WSL (Ubuntu) and it wasn't working. Do you have a separate OS instance running CentOS or you run it in WSL?

**Comment by mbsyaswanth** [comment]
> I'm not sure if this is my setting problem. I modified the 'torchrun-script.py' and then it looks like it works. But I saw a line. ![image](https://private-user-images.githubusercontent.com/16273693/289196293-ee490a66-dea8-4507-a299-ac66462491fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM5NjA0MTgsIm5iZiI6MTcwMzk2MDExOCwicGF0aCI6Ii8xNjI3MzY5My8yODkxOTYyOTMtZWU0OTBhNjYtZGVhOC00NTA3LWEyOTktYWM2NjQ2MjQ5MWZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjMwVDE4MTUxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxZjg2NDBiODY5ZjAxN2QyODg0OWIwMWJhN2ZlMzVhMjFjNDllYWIwMjA0MTBmY2ZiZTdkZjlhZDVhZjYyYWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qFv5RuKzJfu0Lbcb-6q_YVE_Mf8xtT2HPH5XjcPNCUI) So, maybe this model is not working under Windows. Because I use my server PC(CentOS) it is working.

I was also able to get until here after replacing the first line of this file 
`D:\ProgramFiles\miniconda3\Scripts\torchrun-script.py`
to
`#!D:\ProgramFiles\miniconda3\python.exe`

This is basically the installation location of miniconda in my windows conda environment.

**Comment by ZTGeng** [comment]
> I'm not sure if this is my setting problem. I modified the 'torchrun-script.py' and then it looks like it works. But I saw a line. ![image](https://private-user-images.githubusercontent.com/16273693/289196293-ee490a66-dea8-4507-a299-ac66462491fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NjU4MDIsIm5iZiI6MTcxMTY2NTUwMiwicGF0aCI6Ii8xNjI3MzY5My8yODkxOTYyOTMtZWU0OTBhNjYtZGVhOC00NTA3LWEyOTktYWM2NjQ2MjQ5MWZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDIyMzgyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3NzY5ODUxNWJjZDgxOWU4MzRkYTJlOTVjM2E1ZGI1Y2I0NjIxOWUyNTFlN2VjOTViYWY2ODM1MjI0NDY0NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Tl3Om9-ghgOzC4K6BQCfNTuhPDAB91ExycJbUqH-rk0) So, maybe this model is not working under Windows. Because I use my server PC(CentOS) it is working.

The problem is in this line:
\llama\generation.py
line 85: torch.distributed.init_process_group("nccl")

**Simply change "nccl" in this line to "gloo"** will walkaround the issue, per ChatGPT's suggestion.

This argument specifies the backend for distributed communication between GPUs. "NCCL" is a backend supported only on Linux, whereas "GLOO" is suitable for Windows. It's important to note that this isn't a real solution to the underlying issue, and I'm not entirely clear on the differences between these two backends. However, I discovered that my PC has only one Nvidia graphics card, so I don't believe I require distributed communication. Please evaluate this workaround to determine if it's applicable in your situation.

**Comment by ZTGeng** [comment]
By the way you will still get the "Redirects are currently not supported in Windows or MacOs." warning, but looks like it's not a fatal issue.
